pkg/collect/host_dns.go

```diff
@@ -0,0 +1,149 @@
+package collect
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/miekg/dns"
+	"github.com/pkg/errors"
+	troubleshootv1beta2 "github.com/replicatedhq/troubleshoot/pkg/apis/troubleshoot/v1beta2"
+	"k8s.io/klog/v2"
+)
+
+type CollectHostDNS struct {
+	hostCollector *troubleshootv1beta2.HostDNS
+	BundlePath    string
+}
+
+type DNSResult struct {
+	Query DNSQuery `json:"query"`
+}
+
+type DNSQuery map[string][]DNSEntry
+
+type DNSEntry struct {
+	Server string `json:"server"`
+	Name   string `json:"name"`
+	Answer string `json:"answer"`
+}
+
+const (
+	HostDNSPath = "host-collectors/dns/"
+	resolvConf  = "/etc/resolv.conf"
+)
+
+func (c *CollectHostDNS) Title() string {
+	return hostCollectorTitleOrDefault(c.hostCollector.HostCollectorMeta, "dns")
+}
+
+func (c *CollectHostDNS) IsExcluded() (bool, error) {
+	return isExcluded(c.hostCollector.Exclude)
+}
+
+func (c *CollectHostDNS) Collect(progressChan chan<- interface{}) (map[string][]byte, error) {
+
+	names := c.hostCollector.Hostnames
+	if len(names) == 0 {
+		// if no names are provided, query a wilcard to detect wildcard DNS if any
+		names = append(names, "*")
+	}
+	output := NewResult()
+
+	// first, get DNS config from /etc/resolv.conf
+	dnsConfig, err := getDNSConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read DNS resolve config")
+	}
+
+	// query DNS for each name
+	dnsEntries := make(map[string][]DNSEntry)
+	for _, name := range names {
+		entries, err := resolveName(name, dnsConfig)
+		if err != nil {
+			klog.V(2).Infof("Failed to resolve name %s: %v", name, err)
+		}
+		dnsEntries[name] = entries
+	}
+	dnsResult := DNSResult{Query: dnsEntries}
+
+	// convert dnsResult to a JSON string
+	dnsResultJSON, err := json.MarshalIndent(dnsResult, "", "  ")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal DNS query result to JSON")
+	}
+
+	outputFile := filepath.Join(HostDNSPath, "results.json")
+	output.SaveResult(c.BundlePath, outputFile, bytes.NewBuffer(dnsResultJSON))
+
+	// write /etc/resolv.conf to a file
+	resolvConfData, err := getResolvConf()
+	if err != nil {
+		klog.V(2).Infof("failed to read DNS resolve config: %v", err)
+	} else {
+		outputFile = filepath.Join(HostDNSPath, "resolv.conf")
+		output.SaveResult(c.BundlePath, outputFile, bytes.NewBuffer(resolvConfData))
+	}
+
+	return output, nil
+}
+
+func getDNSConfig() (*dns.ClientConfig, error) {
+	file, err := os.Open(resolvConf)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	config, err := dns.ClientConfigFromFile(file.Name())
```

Comment:
Should we have an option on the collector to specify ndots, and then if its set update the DNS config here to set that value?

https://github.com/miekg/dns/blob/master/clientconfig.go#L12-L19

Reply:
ndots is a good example, but anywhere someone might want to override the config from the collector could be useful to expose.

Reply:
Looking further, since you're iterating and appending every search domain found on the host to see if one of them resolves something we suspect it shouldn't, then the ndots stuff probably isn't needed? So whether you decide to add that or not completely up to you at this stage.